In [3]:
import pandas as pd

In [ ]:
df = pd.read_csv('companies_10k_reports.csv', usecols=range(7))

for col in df.columns[2:]:
    df[col] = df[col].str.replace(',', '').astype(int)

Company  Fiscal Year  Total Revenue (in millions USD)  \
0  Microsoft         2024                           245122   
1  Microsoft         2023                           211915   
2  Microsoft         2022                           198270   
3      Tesla         2024                            97690   
4      Tesla         2023                            96773   

   Net Income (in millions USD)  Total Assets (in millions USD)  \
0                         88136                          512163   
1                         72361                          411976   
2                         72738                          364840   
3                          7153                          122070   
4                         14974                          106618   

   Total Liabilities (in millions USD)  Operating Cash Flow (in millions USD)  
0                               243686                                 118548  
1                               205753                                  87582  
2                               198298                                  89035  
3                                48390                                  14923  
4                                43009                                  13256

In [22]:
df = df.sort_values(by=['Company', 'Fiscal Year'])
df['Revenue Growth (%)'] = df.groupby(['Company'])['Total Revenue (in millions USD)'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby(['Company'])['Net Income (in millions USD)'].pct_change() * 100
df.head(10)

Company  Fiscal Year  Total Revenue (in millions USD)  \
8      Apple         2022                           394328   
7      Apple         2023                           383285   
6      Apple         2024                           391035   
2  Microsoft         2022                           198270   
1  Microsoft         2023                           211915   
0  Microsoft         2024                           245122   
5      Tesla         2022                            81462   
4      Tesla         2023                            96773   
3      Tesla         2024                            97690   

   Net Income (in millions USD)  Total Assets (in millions USD)  \
8                         99803                          352755   
7                         96995                          352583   
6                         93736                          364980   
2                         72738                          364840   
1                         72361                          411976   
0                         88136                          512163   
5                         12587                           82338   
4                         14974                          106618   
3                          7153                          122070   

   Total Liabilities (in millions USD)  Operating Cash Flow (in millions USD)  \
8                               302083                                 122151   
7                               290437                                 110543   
6                               308030                                 118254   
2                               198298                                  89035   
1                               205753                                  87582   
0                               243686                                 118548   
5                                36440                                  14724   
4                                43009                                  13256   
3                                48390                                  14923   

   Revenue Growth (%)  Net Income Growth (%)  
8                 NaN                    NaN  
7           -2.800461              -2.813543  
6            2.021994              -3.359967  
2                 NaN                    NaN  
1            6.882030              -0.518299  
0           15.669962              21.800417  
5                 NaN                    NaN  
4           18.795267              18.964010  
3            0.947578             -52.230533

In [ ]:
summary_stats = df.groupby('Company')[[
    'Total Revenue (in millions USD)',
    'Net Income (in millions USD)',
    'Operating Cash Flow (in millions USD)',
    'Revenue Growth (%)',
    'Net Income Growth (%)'
]].agg(['mean', 'std', 'min', 'max'])

summary_stats.head()

Total Revenue (in millions USD)                                \
                                     mean           std     min     max   
Company                                                                   
Apple                       389549.333333   5669.423810  383285  394328   
Microsoft                   218435.666667  24097.028786  198270  245122   
Tesla                        91975.000000   9116.062692   81462   97690   

          Net Income (in millions USD)                             \
                                  mean          std    min    max   
Company                                                             
Apple                     96844.666667  3036.292531  93736  99803   
Microsoft                 77745.000000  9000.844016  72361  88136   
Tesla                     11571.333333  4008.203380   7153  14974   

          Operating Cash Flow (in millions USD)                                \
                                           mean           std     min     max   
Company                                                                         
Apple                             116982.666667   5907.506440  110543  122151   
Microsoft                          98388.333333  17473.892592   87582  118548   
Tesla                              14301.000000    910.449889   13256   14923   

          Revenue Growth (%)                                  \
                        mean        std       min        max   
Company                                                        
Apple              -0.389233   3.409990 -2.800461   2.021994   
Microsoft          11.275996   6.214007  6.882030  15.669962   
Tesla               9.871422  12.620221  0.947578  18.795267   

          Net Income Growth (%)                                   
                           mean        std        min        max  
Company                                                           
Apple                 -3.086755   0.386380  -3.359967  -2.813543  
Microsoft             10.641059  15.781715  -0.518299  21.800417  
Tesla                -16.633261  50.342144 -52.230533  18.964010

In [21]:
trend_by_year = df.groupby('Fiscal Year')[[
    'Total Revenue (in millions USD)',
    'Net Income (in millions USD)'
]].mean().reset_index()
trend_by_year.head()

Fiscal Year  Total Revenue (in millions USD)  Net Income (in millions USD)
0         2022                    224686.666667                  61709.333333
1         2023                    230657.666667                  61443.333333
2         2024                    244615.666667                  63008.333333

## Summary of Financial Performance Trends

- **Apple** showed a slight revenue dip in 2023 followed by a modest recovery in 2024. However, its net income declined both years, suggesting increasing costs or other profitability challenges.
- **Microsoft** demonstrated strong, accelerating revenue growth, especially in 2024 (+15.7%). Net income growth also surged in 2024 (+21.8%), indicating operational efficiency or favorable margins.
- **Tesla** had strong revenue growth in 2023 (+18.8%) but saw a massive net income drop in 2024 (-52.2%), possibly due to high expenditures or reduced margins despite rising revenue.

### Key Takeaways:
- Microsoft stands out as the most consistently growing company in both revenue and income.
- Tesla’s 2024 income collapse needs further analysis—cost spikes or margin shrinkage could be drivers.
- Apple’s plateauing metrics suggest a maturing or saturated market presence.

These trends highlight the importance of not just top-line growth but bottom-line health, especially when analyzing long-term sustainability.